src/platform/Linux/CHIPBluezHelper.cpp

```diff
@@ -1710,6 +1667,101 @@ CHIP_ERROR InitBluezBleLayer(bool aIsCentral, char * apBleAddr, BLEAdvConfig & a
     return err;
 }
 
+// StartDiscovery callbacks
+
+using DiscoveryTaskArg = std::pair<BluezEndpoint *, BluezDiscoveryRequest>;
+
+void StartDiscoveryDone(GObject * aObject, GAsyncResult * aResult, gpointer apEndpoint)
+{
+    BluezAdapter1 * adapter = BLUEZ_ADAPTER1(aObject);
+    GError * error          = nullptr;
+    gboolean success        = bluez_adapter1_call_start_discovery_finish(adapter, aResult, &error);
+
+    VerifyOrExit(success == TRUE, ChipLogProgress(DeviceLayer, "FAIL: StartDiscovery : %s", error->message));
+    ChipLogProgress(DeviceLayer, "StartDiscovery complete");
+
+exit:
+    if (error != nullptr)
+        g_error_free(error);
+}
+
+static gboolean StartDiscoveryImpl(void * apDiscoveryTaskArg)
+{
+    DiscoveryTaskArg * taskArg = static_cast<DiscoveryTaskArg *>(apDiscoveryTaskArg);
+    BluezEndpoint * endpoint;
+
+    VerifyOrExit(taskArg != nullptr, ChipLogProgress(DeviceLayer, "taskArg is NULL in %s", __func__));
+    endpoint = taskArg->first;
+
+    VerifyOrExit(endpoint != nullptr, ChipLogProgress(DeviceLayer, "endpoint is NULL in %s", __func__));
+    VerifyOrExit(endpoint->mpAdapter != nullptr, ChipLogProgress(DeviceLayer, "mpAdapter is NULL in %s", __func__));
+
+    endpoint->mDiscoveryRequest = taskArg->second;
+    delete taskArg;
+
+    bluez_adapter1_call_start_discovery(endpoint->mpAdapter, nullptr, StartDiscoveryDone, endpoint);
+
+exit:
+    return G_SOURCE_REMOVE;
+}
+
+CHIP_ERROR StartDiscovery(BluezEndpoint * apEndpoint, const BluezDiscoveryRequest request)
+{
+    DiscoveryTaskArg * const taskArg = new DiscoveryTaskArg(apEndpoint, request);
+
+    if (!BluezRunOnBluezThread(StartDiscoveryImpl, taskArg))
+    {
+        ChipLogError(Ble, "Failed to schedule StartDiscoveryImpl() on CHIPoBluez thread");
+        delete taskArg;
+        return CHIP_ERROR_INCORRECT_STATE;
```

Comment:
Prefer to maintain a single point of exit within the method/function. I would define the `error` to `CHIP_NO_ERROR`  at the beginning.

```suggestion
        ExitNow(error = CHIP_ERROR_INCORRECT_STATE);
```

Reply:
Ok, but why is it better in case no deinitialization code is needed? For me calling "return" explicitly is more readable since I don't need to scroll down to see if there's anything unusual done in the epilogue.

Reply:
I think there is no way to say which approach is better or right. In my opinion (and it's in many coding standards), it's much safer to have a single return statement per function or method at the end of it. Mainly to reduce risk when a function is modified later on etc..